elasticapm/contrib/serverless/azure.py

```diff
@@ -0,0 +1,218 @@
+
+_AnnotatedFunctionT = TypeVar("_AnnotatedFunctionT")
+
+_cold_start_lock = threading.Lock()
+COLD_START = True
+
+
+class AzureFunctionsClient(Client):
+    def get_service_info(self):
+        service_info = super().get_service_info()
+        service_info["framework"] = {
+            "name": "Azure Functions",
+            "version": os.environ.get("FUNCTIONS_EXTENSION_VERSION"),
+        }
+        service_info["runtime"] = {
+            "name": os.environ.get("FUNCTIONS_WORKER_RUNTIME"),
+            "version": os.environ.get("FUNCTIONS_WORKER_RUNTIME_VERSION"),
+        }
+        service_info["node"] = {"configured_name": os.environ.get("WEBSITE_INSTANCE_ID")}
+        return service_info
+
+    def get_cloud_info(self):
+        cloud_info = super().get_cloud_info()
+        cloud_info.update(
+            {"provider": "azure", "region": os.environ.get("REGION_NAME"), "service": {"name": "functions"}}
+        )
+        account_id = get_account_id()
+        if account_id:
+            cloud_info["account"] = {"id": account_id}
+        if "WEBSITE_SITE_NAME" in os.environ:
+            cloud_info["instance"] = {"name": os.environ["WEBSITE_SITE_NAME"]}
+        if "WEBSITE_RESOURCE_GROUP" in os.environ:
+            cloud_info["project"] = {"name": os.environ["WEBSITE_RESOURCE_GROUP"]}
+        return cloud_info
+
+
+class ElasticAPMExtension(AppExtensionBase):
+    functions = {}
+    client = None
+
+    @classmethod
+    def init(cls):
+        elasticapm.instrument()
+
+    @classmethod
+    def configure(cls, client_class=AzureFunctionsClient, **kwargs):
+        client = elasticapm.get_client()
+        if not client:
+            kwargs["metrics_interval"] = "0ms"
+            kwargs["breakdown_metrics"] = "false"
+            if "metric_sets" not in kwargs and "ELASTIC_APM_METRICS_SETS" not in os.environ:
+                # Allow users to override metrics sets
+                kwargs["metrics_sets"] = []
+            kwargs["central_config"] = "false"
+            kwargs["cloud_provider"] = "none"
+            kwargs["framework_name"] = "Azure Functions"
+            if (
+                "service_name" not in kwargs
+                and "ELASTIC_APM_SERVICE_NAME" not in os.environ
+                and "WEBSITE_SITE_NAME" in os.environ
+            ):
+                kwargs["service_name"] = os.environ["WEBSITE_SITE_NAME"]
+            if (
+                "environment" not in kwargs
+                and "ELASTIC_APM_ENVIRONMENT" not in os.environ
+                and "AZURE_FUNCTIONS_ENVIRONMENT" in os.environ
+            ):
+                kwargs["environment"] = os.environ["AZURE_FUNCTIONS_ENVIRONMENT"]
+            client = AzureFunctionsClient(**kwargs)
+        cls.client = client
+
+    @classmethod
+    def post_function_load_app_level(cls, function_name: str, function_directory: str, *args, **kwargs):
+        with codecs.open(os.path.join(function_directory, "function.json")) as f:
```

Comment:
Any particular reason you're using `codecs.open` instead of just `open`? My understanding is that `codecs.open` is obsolete in py3.x but I could have missed something.